src/legacy/core_plugins/kibana/public/dashboard/saved_dashboard/saved_dashboard.js

```diff
@@ -103,6 +106,34 @@ module.factory('SavedDashboard', function (Private, config, i18n) {
 
   SavedDashboard.searchsource = true;
 
+  SavedDashboard.extractReferences = function (source) {
+    // Extract references on save here
+    const references = _.clone(source.references) || {};
+    const panels = JSON.parse(source.panelsJSON);
+    panels.forEach((panel, i) => {
+      panel.panelReference = `panel_${i}`;
+      references[`panel_${i}`] = pick(panel, 'type', 'id');
+      delete panel.type;
+      delete panel.id;
+    });
+    return assign({}, source, {
+      references,
+      panelsJSON: JSON.stringify(panels)
+    });
+  };
+
+  SavedDashboard.injectReferences = function (references) {
+    // Inject references on read here
+    const panels = JSON.parse(this.panelsJSON);
+    panels.forEach((panel) => {
+      const reference = references[panel.panelReference];
+      panel.id = reference.id;
+      panel.type = reference.type;
+      delete panel.panelReference;
+    });
+    this.panelsJSON = JSON.stringify(panels);
```

Comment:
Probably have a way that doesn't modify `this` and works with arguments instead.